multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/UploadAppStep.java

```diff
@@ -88,20 +89,36 @@ public StepPhase executeAsyncStep(ProcessContext context) throws FileStorageExce
             return StepPhase.POLL;
         }
 
-        Optional<CloudPackage> latestUnusedPackage = cloudPackagesGetter.getLatestUnusedPackage(client, cloudApp.getGuid());
-        if (latestUnusedPackage.isPresent() && isCloudPackageInValidState(latestUnusedPackage.get())) {
-            return useLatestPackage(context, latestUnusedPackage.get());
+        Optional<CloudPackage> mostRecentPackage = cloudPackagesGetter.getMostRecentAppPackage(client, cloudApp.getGuid());
+        if (mostRecentPackage.isEmpty()) {
+            proceedWithUpload(context, applicationToProcess, moduleFileName);
+            return StepPhase.POLL;
         }
 
-        if (latestUnusedPackage.isPresent() && !isCloudPackageInValidState(latestUnusedPackage.get())
-            || !isAppStagedCorrectly(context, cloudApp)) {
+        CloudPackage latestPackage = mostRecentPackage.get();
+        Optional<CloudPackage> currentPackage = cloudPackagesGetter.getAppPackage(client, cloudApp.getGuid());
+        if (currentPackage.isEmpty() && isPackageInValidState(latestPackage)) {
+            return useLatestPackage(context, latestPackage);
+        }
+
+        if (currentPackage.isEmpty() || !isAppStagedCorrectly(context, cloudApp)) {
             proceedWithUpload(context, applicationToProcess, moduleFileName);
             return StepPhase.POLL;
         }
+
+        if (isPackageInValidState(latestPackage) && (context.getVariable(Variables.APP_STACK_CHANGED) ||
```

Comment:
You can use Variables.VCAP_APP_PROPERTIES_CHANGED instead of Variables.APP_STACK_CHANGED